src/pint/simulation.py

```diff
@@ -474,36 +435,175 @@ def calculate_random_models(
-
     # scale by fac
     mean_vector = mean_vector * fac
     scaled_cov_matrix = ((cov_matrix.matrix * fac).T * fac).T
     random_models = []
-    for imodel in range(Nmodels):
+    for _ in range(Nmodels):
+        f_rand = deepcopy(fitter)
         # create a set of randomized parameters based on mean vector and covariance matrix
-        rparams_num = np.random.multivariate_normal(mean_vector, scaled_cov_matrix)
-        # scale params back to real units
-        for j in range(len(mean_vector)):
-            rparams_num[j] /= fac[j]
-        rparams = OrderedDict(zip(param_names, rparams_num))
-        f_rand.set_params(rparams)
-        phase = f_rand.model.phase(toas, abs_phase=True)
-        phases_i[imodel] = phase.int
-        phases_f[imodel] = phase.frac
-        r = pint.residuals.Residuals(toas, f_rand.model)
-        freqs[imodel] = r.get_PSR_freq(calctype="taylor")
-        if keep_models:
-            random_models.append(f_rand.model)
-            f_rand = deepcopy(fitter)
-    phases = phases_i + phases_f
-    phases0 = fitter.model.phase(toas, abs_phase=True)
-    dphase = phases - (phases0.int + phases0.frac)
-
-    if return_time:
-        dphase /= freqs
-
-    if keep_models:
-        return dphase, random_models
-    else:
-        return dphase
+        # dividing by fac brings us back to real units
+        rparams = np.random.multivariate_normal(mean_vector, scaled_cov_matrix) / fac
+        f_rand.set_params(dict(zip(param_names, rparams)))
+        random_models.append(f_rand.model)
+    return random_models
+
+
+def compute_random_model_resids_exact(fitter, models, toas, return_time=False):
+    """
+    Calculate residuals from random models exactly
+
+    Parameters
+    ----------
+    fitter : pint.fitter.Fitter
+        current post-fit fitter object instance containing the reference model
+    models : list, pint.models.timing_model.TimingModel
+        list of the random models to compute residuals
+    toas : pint.toa.TOAs
+        TOAs to calculate residuals from
+    return_time : bool, default=False
+        return the residuals in time rather than in phase
+
+    Returns
+    -------
+    dphase : np.ndarray
+        phase difference with respect to reference model, size is [len(models), len(toas)]
+
+    Note
+    ----
+    This function is used by :func:`~pint.simulation.calculate_random_models`
+    """
+    resids = np.zeros((len(models), toas.ntoas), dtype=np.float)
+    # These are the reference residuals from the reference model
+    r0 = pr.Residuals(toas, fitter.model, subtract_mean=False)
+    for ii, model in enumerate(models):
+        rn = pr.Residuals(toas, model, subtract_mean=False)
+        resids[ii] = (
+            rn.time_resids - r0.time_resids
+            if return_time
+            else rn.phase_resids - r0.phase_resids
+        )
+    return resids << u.s if return_time else resids
+
+
+def compute_random_model_resids_fast(fitter, models, toas, return_time=False):
```

Comment:
The design matrix can also be directly used to compute one-sigma "error bands", with a small additional amount of linear algebra. Such error bands don't fully capture the shape of deviations but can be a very tidy way to summarize the uncertainty in a model. Would it be good to have a function to compute these and a plotting mode in pintk to show them?

Reply:
This is basically exactly how/why we use the random models now in `pintk`.  But I definitely think it might be useful to be able to do this computation, for instance, at the time of a non-phase-connected TOA in the future to see if it is likely phase-connected or not.  We use random models now for that, but this is probably a much more "surgical" option. Do you know of a ref for the linear algebra to do this?